File: 1408-find-the-smallest-divisor-given-a-threshold/1408-find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:

    int sumByD(vector<int> arr, int divisor) {
        int n = arr.size();
        int sum = 0;
        for(int i = 0; i<n; i++) {
            sum += ceil((double)(arr[i])/(double)(divisor));
        }
        return sum;
    }
    int smallestDivisor(vector<int>& nums, int threshold) {
        int l = 1, h = *max_element(nums.begin(), nums.end()), ans = -1;
        while(l<=h) {
            int m = (l+h)/2;
            if(sumByD(nums, m) <= threshold) {
                ans = m;
                h = m-1;
            }
            else l = m+1;
        }
        return ans;
    }
};

File: 1408-find-the-smallest-divisor-given-a-threshold/1408-find-the-smallest-divisor-given-a-threshold.java
class Solution {
    public int smallestDivisor(int[] nums, int threshold) {
        int start=1;
        int end=Arrays.stream(nums).max().getAsInt();
        int smallestInt=0;
        while(start<=end){
            int mid=start+(end-start)/2;
            if(findSum(nums,mid)>threshold){
                start=mid+1;
            }else{
                smallestInt=mid;
                end=mid-1;

            }
        }
        return smallestInt;
        
    }
    public static int findSum(int[] nums,int div){
        int sum=0;
        for(int n:nums){
         sum+=Math.ceil((double)n/(double)div);
        }
        return sum;
    }

}